File: core/DungeonGame/DungeonGame.java
package com.dbc;

import java.util.Arrays;

public class DungeonGame {
    public int calculateMinimumHP(int[][] dungeon) {
        int rows = dungeon.length, cols = dungeon[0].length;
        int[][] dp = new int[rows + 1][cols + 1];
        for (int i = 0; i <= rows; i++) {
            Arrays.fill(dp[i], Integer.MAX_VALUE);
        }
        dp[rows][cols - 1] = dp[rows - 1][cols] = 1;
        for (int row = rows - 1; row >= 0; row--) {
            for (int col = cols - 1; col >=0; col--) {
                int min = Math.min(dp[row + 1][col], dp[row][col + 1]);
                dp[row][col] = Math.max(min - dungeon[row][col], 1);
            }
        }
        return dp[0][0];
    }
}

File: core/DungeonGame/DungeonGame.py
from typing import List


def calculateMinimumHP(dungeon: List[List[int]]) -> int:
    res, rows, cols = 0, len(dungeon), len(dungeon[0])
    big = 10**9
    dp = [[big] * (cols + 1) for _ in range(rows + 1)]
    dp[rows][cols - 1] = dp[rows - 1][cols] = 1
    for i in range(rows - 1, -1, -1):
        for j in range(cols - 1, -1, -1):
            minn = min(dp[i + 1][j], dp[i][j + 1])
            dp[i][j] = max(minn - dungeon[i][j], 1)

    return dp[0][0];

if __name__ == '__main__':
    print(calculateMinimumHP([[0,5],[-2,-3]]))



